File: reference/RidmikParser.java
package net.hasnath.ridmikparser;
public class RidmikParser {
    private net.hasnath.ridmikparser.BanglaUnicode unicode;

    public RidmikParser()
    {
        this.unicode = new net.hasnath.ridmikparser.BanglaUnicode();
        return;
    }

    public static void main(String[] p5)
    {
        net.hasnath.ridmikparser.RidmikParser v1_1 = new net.hasnath.ridmikparser.RidmikParser();
        java.util.Scanner v2_1 = new java.util.Scanner(System.in);
        System.out.println("Enter:");
        while (v2_1.hasNext()) {
            System.out.println(v1_1.toBangla(v2_1.nextLine()));
        }
        return;
    }

    boolean dualSitsUnder(char p6, char p7, char p8, char p9)
    {
        String v2_0 = 1;
        if ((p7 != 114) || (p6 != 114)) {
            if (p7 != 114) {
                String v0 = this.unicode.getDjkt(p8, p9);
                if ((v0 == null) || (!this.isCharInString(p7, v0))) {
                    String v1 = this.unicode.getDjktt(p8, p9);
                    if (v1 == null) {
                        v2_0 = 0;
                    } else {
                        v2_0 = v1.contains(new StringBuilder(String.valueOf(Character.toString(p6))).append(Character.toString(p7)).toString());
                    }
                }
            } else {
                v2_0 = 0;
            }
        }
        return v2_0;
    }

    boolean isCharInString(char p3, String p4)
    {
        int v0_1;
        if (p4.indexOf(p3) != -1) {
            v0_1 = 1;
        } else {
            v0_1 = 0;
        }
        return v0_1;
    }

    boolean isConsonant(char p2)
    {
        if ((this.isVowel(p2)) || (!Character.isLetter(p2))) {
            int v0_2 = 0;
        } else {
            v0_2 = 1;
        }
        return v0_2;
    }

    boolean isVowel(char p3)
    {
        int v0_2;
        if ("AEIOUaeiou".indexOf(p3) != -1) {
            v0_2 = 1;
        } else {
            v0_2 = 0;
        }
        return v0_2;
    }

    boolean notJukta(char p6, char p7, char p8, char p9)
    {
        int v1 = 0;
        if ((p7 != 110) || ((p8 != 103) || (p9 != 114))) {
            if ((p9 != 114) && ((p9 != 122) && (p9 != 119))) {
                String v0_0 = this.unicode.getDualJkt(p7, p8);
                if (v0_0 == null) {
                    String v0_1 = this.unicode.getJkt(p8);
                    if (v0_1 == null) {
                        v1 = 1;
                    } else {
                        if (!this.isCharInString(p9, v0_1)) {
                            v1 = 1;
                        }
                    }
                } else {
                    if (!this.isCharInString(p9, v0_0)) {
                        v1 = 1;
                    }
                }
            }
        } else {
            v1 = 1;
        }
        return v1;
    }

    public String toBangla(String p19)
    {
        StringBuilder v10_1 = new StringBuilder();
        int v1 = 0;
        int v9 = 0;
        int v12 = 0;
        int v5 = 0;
        int v8 = 0;
        int v7 = 0;
        char[] v2 = p19.toCharArray();
        int v14 = v2.length;
        int v13_0 = 0;
        while (v13_0 < v14) {
            int v6 = v2[v13_0];
            if (((v6 >= 97) && (v6 <= 122)) || (((v6 >= 65) && (v6 <= 90)) || ((v6 >= 48) && (v6 <= 57)))) {
                if ((v6 == 65) || ((v6 == 66) || ((v6 == 67) || ((v6 == 69) || ((v6 == 70) || ((v6 == 80) || (v6 == 88))))))) {
                    v6 = Character.toLowerCase(v6);
                }
                if ((v6 == 75) || ((v6 == 76) || ((v6 == 77) || ((v6 == 86) || ((v6 == 89) || ((v6 == 87) || (v6 == 81))))))) {
                    v6 = Character.toLowerCase(v6);
                }
                if ((v6 == 72) && (v1 != 84)) {
                    v6 = 104;
                }
                if (((v1 == 0) || (this.isVowel(v1))) && (v6 == 119)) {
                    v6 = 79;
                }
                if (this.isVowel(v6)) {
                    if ((v1 != 114) || ((v9 != 114) || (v6 != 105))) {
                        String v3_1;
                        if (v9 == 0) {
                            v3_1 = this.unicode.getDual(v6, v1);
                        } else {
                            v3_1 = this.unicode.getDualKar(v6, v1);
                        }
                        if (v3_1 == null) {
                            if ((v6 != 111) || (v1 == 0)) {
                                if ((!this.isVowel(v1)) && (v1 != 0)) {
                                    v10_1.append(this.unicode.getKar(v6));
                                } else {
                                    if ((v6 != 97) || (v1 == 0)) {
                                        v10_1.append(this.unicode.get(v6));
                                    } else {
                                        v10_1.append(this.unicode.get(121)).append(this.unicode.getKar(97));
                                    }
                                }
                            } else {
                                if (!this.isVowel(v1)) {
                                    v12 = v9;
                                    v9 = v1;
                                    v1 = v6;
                                    v13_0++;
                                } else {
                                    v10_1.append(this.unicode.get(79));
                                }
                            }
                        } else {
                            if (v1 != 111) {
                                v10_1.delete((v10_1.length() - 1), v10_1.length());
                            }
                            if (!this.isVowel(v9)) {
                                v10_1.append(v3_1);
                            } else {
                                v10_1.append(this.unicode.get(v1)).append(this.unicode.get(v6));
                            }
                        }
                    } else {
                        if (v12 != 0) {
                            v10_1.delete((v10_1.length() - 3), v10_1.length());
                            v10_1.append("\u09c3");
                        } else {
                            v10_1.delete((v10_1.length() - 2), v10_1.length());
                            v10_1.append("\u098b");
                        }
                        v1 = 105;
                    }
                }
                if ((v6 == 121) || ((v6 == 90) || (v6 == 114))) {
                    v5 = 0;
                }
                if ((v5 == 0) || (this.unicode.getDual(v6, v1) != null)) {
                    int v11 = 0;
                } else {
                    v11 = 1;
                }
                if ((!this.isConsonant(v6)) || ((!this.isConsonant(v1)) || (v11 != 0))) {
                    if (this.isConsonant(v6)) {
                        v7 = 0;
                        if ((this.isVowel(v1)) && (v6 == 90)) {
                            v10_1.append("\u09cd");
                        }
                        if ((v1 == 0) && (v6 == 120)) {
                            v10_1.append(this.unicode.get(101));
                        }
                        v8 = v5;
                        v5 = 0;
                        if ((v6 == 119) && ((this.isConsonant(v1)) && (this.isConsonant(v9)))) {
                            v10_1.append("\u09cd");
                            v8 = 0;
                            v5 = 1;
                        }
                        if ((v12 == 107) && ((v9 == 83) && ((v1 == 104) && ((v6 == 78) || (v6 == 109))))) {
                            v10_1.append("\u09cd");
                            v8 = 0;
                            v5 = 1;
                        }
                        v10_1.append(this.unicode.get(v6));
                    }
                } else {
                    if (((v6 == 121) || (v6 == 90)) && ((v6 != 121) || ((v1 != 113) || (v9 != 113)))) {
                        v6 = 122;
                    }
                    if ((v9 == 107) && ((v1 == 107) && (v6 == 104))) {
                        v1 = 83;
                    }
                    String v3_0 = this.unicode.getDual(v6, v1);
                    if ((v3_0 == null) || (v7 != 0)) {
                        v7 = 0;
                        v8 = v5;
                        v5 = 0;
                        if ((v9 == 114) || ((v1 != 114) || (v6 != 122))) {
                            if (((v1 != 114) || (v9 == 114)) && ((v1 != 114) || ((v9 != 114) || (!this.isConsonant(v12))))) {
                                if ((v1 != 114) || ((v9 != 114) || ((!this.isVowel(v12)) && (v12 != 0)))) {
                                    if (!this.notJukta(v12, v9, v1, v6)) {
                                        v10_1.append("\u09cd");
                                        v5 = 1;
                                    }
                                } else {
                                    v10_1.delete((v10_1.length() - 1), v10_1.length());
                                    v10_1.append("\u09cd");
                                }
                            }
                        } else {
                            v10_1.append("\u200d\u09cd");
                        }
                        v10_1.append(this.unicode.get(v6));
                    } else {
                        v7 = 1;
                        if ((v12 == 103) && ((v9 == 107) && ((v1 == 83) && (v6 == 104)))) {
                            v5 = 0;
                            v8 = 0;
                        }
                        if ((this.isVowel(v9)) || ((v9 == 0) || (v8 != 0))) {
                            int v4 = 1;
                        } else {
                            v4 = 0;
                        }
                        if ((!this.dualSitsUnder(v12, v9, v1, v6)) || (v4 != 0)) {
                            if (v5 == 0) {
                                v10_1.delete((v10_1.length() - 1), v10_1.length());
                            } else {
                                v10_1.delete((v10_1.length() - 2), v10_1.length());
                            }
                            v10_1.append(v3_0);
                            v8 = v5;
                            v5 = 0;
                        } else {
                            v10_1.delete((v10_1.length() - 1), v10_1.length());
                            if ((v9 == 114) && (v12 == 114)) {
                                v10_1.delete((v10_1.length() - 1), v10_1.length());
                            }
                            if ((v5 == 0) && ((v9 != 0) && (!this.isVowel(v9)))) {
                                v10_1.append("\u09cd");
                            }
                            v10_1.append(v3_0);
                            v8 = v5;
                            v5 = 1;
                        }
                    }
                }
                v12 = v9;
                v9 = v1;
                v1 = v6;
            } else {
                v10_1.append(v6);
                v1 = 0;
            }
        }
        return v10_1.toString();
    }
}

File: app/src/main/java/com/kinetic/keyboard/input/RidmikParser.kt
package com.kinetic.keyboard.input

/**
 * Banglish → Bangla phonetic transliterator, ported from the recovered reference
 * (reference/RidmikParser.java, decompiled; SPEC.md P3.2).
 *
 * DELIBERATELY LITERAL. The decompiled source is not directly compilable (boolean/int confusion,
 * lost loop structure), so this is a faithful reconstruction of its logic; the golden corpus in
 * RidmikParserTest locks the behavior (each expectation hand-traced through the original control
 * flow). Readability refactor is P3.6 — until then, keep the original variable roles:
 *
 *   v1  = previous char        v9 = char before that     v12 = char before that
 *   v5  = "conjunct just formed" flag                    v8  = saved v5
 *   v7  = "second char of a two-char consonant was consumed" flag
 *
 * Sliding-window rules of note:
 *   consonant+consonant auto-joins with ্ when the jkt tables allow (rokto → রক্ত);
 *   two-char consonants (kh gh ch …) replace their first char (kokhono → কখন);
 *   'o' after a consonant is the inherent vowel (silent); rr+consonant → reph (korrmo → কর্ম);
 *   rri → ঋ/ৃ; kkh → ক্ষ; word-initial x → এক্স; w is ব-phala inside a cluster.
 */
class RidmikParser {

    private val u = BanglaUnicodeTable()

    private fun isVowel(cp: Int): Boolean = cp > 0 && "AEIOUaeiou".indexOf(cp.toChar()) != -1

    private fun isConsonant(cp: Int): Boolean =
        cp > 0 && !isVowel(cp) && Character.isLetter(cp)

    private fun isCharInString(cp: Int, s: String): Boolean = s.indexOf(cp.toChar()) != -1

    /** Whether current char does NOT auto-join the previous consonant with ্. */
    private fun notJukta(v12: Int, v9: Int, v1: Int, v6: Int): Boolean {
        if (v9 == 'n'.code && v1 == 'g'.code && v6 == 'r'.code) return true
        if (v6 == 'r'.code || v6 == 'z'.code || v6 == 'w'.code) return false
        val dual = u.getDualJkt(v9, v1)
        return if (dual != null) !isCharInString(v6, dual)
        else {
            val single = u.getJkt(v1)
            single == null || !isCharInString(v6, single)
        }
    }

    /** Whether a two-char consonant (v1+v6) sits under the consonant(s) before it. */
    private fun dualSitsUnder(v12: Int, v9: Int, v1: Int, v6: Int): Boolean {
        if (v9 == 'r'.code && v12 == 'r'.code) return true
        if (v9 == 'r'.code) return false
        val d = u.getDjkt(v1, v6)
        if (d != null && isCharInString(v9, d)) return true
        val dd = u.getDjktt(v1, v6)
        return dd != null && dd.contains("${v12.toChar()}${v9.toChar()}")
    }

    fun toBangla(input: String): String {
        val sb = StringBuilder()
        var v1 = 0; var v9 = 0; var v12 = 0
        var v5 = 0; var v8 = 0; var v7 = 0

        for (ch in input) {
            var v6 = ch.code

            // Non-alphanumeric: pass through and reset context.
            if (!(v6 in 97..122 || v6 in 65..90 || v6 in 48..57)) {
                sb.append(ch)
                v1 = 0
                continue
            }

            // Fold uppercase letters whose case carries no meaning.
            when (v6) {
                'A'.code, 'B'.code, 'C'.code, 'E'.code, 'F'.code, 'P'.code, 'X'.code,
                'K'.code, 'L'.code, 'M'.code, 'V'.code, 'Y'.code, 'W'.code, 'Q'.code,
                -> v6 = Character.toLowerCase(v6)
            }
            if (v6 == 'H'.code && v1 != 'T'.code) v6 = 'h'.code
            // w at word start / after a vowel means ও.
            if ((v1 == 0 || isVowel(v1)) && v6 == 'w'.code) v6 = 'O'.code

            if (isVowel(v6)) {
                if (v1 == 'r'.code && v9 == 'r'.code && v6 == 'i'.code) {
                    // rri → ঋ standalone, ৃ after a consonant (the rr chars were already emitted).
                    if (v12 != 0) {
                        sb.delete(sb.length - 3, sb.length)
                        sb.append("ৃ")
                    } else {
                        sb.delete(sb.length - 2, sb.length)
                        sb.append("ঋ")
                    }
                    v1 = 'i'.code
                } else {
                    val v3 = if (v9 == 0) u.getDual(v6, v1) else u.getDualKar(v6, v1)
                    if (v3 == null) {
                        if (v6 == 'o'.code && v1 != 0) {
                            if (!isVowel(v1)) {
                                // Inherent vowel after a consonant: emit nothing, remember 'o'.
                                v12 = v9; v9 = v1; v1 = v6
                                continue
                            } else {
                                sb.append(u.get('O'.code))
                            }
                        } else {
                            if (!isVowel(v1) && v1 != 0) {
                                sb.append(u.getKar(v6)) // dependent sign after consonant
                            } else if (v6 == 'a'.code && v1 != 0) {
                                sb.append(u.get('y'.code)).append(u.getKar('a'.code)) // vowel+a → য়া
                            } else {
                                sb.append(u.get(v6)) // independent vowel
                            }
                        }
                    } else {
                        // Two-char vowel token (OI, OU, oo, …) replaces its first char.
                        if (v1 != 'o'.code) sb.delete(sb.length - 1, sb.length)
                        if (!isVowel(v9)) sb.append(v3)
                        else sb.append(u.get(v1)).append(u.get(v6))
                    }
                }
            }

            if (v6 == 'y'.code || v6 == 'Z'.code || v6 == 'r'.code) v5 = 0
            val v11 = if (v5 == 0 || u.getDual(v6, v1) != null) 0 else 1

            if (!isConsonant(v6) || !isConsonant(v1) || v11 != 0) {
                if (isConsonant(v6)) {
                    v7 = 0
                    if (isVowel(v1) && v6 == 'Z'.code) sb.append("্")
                    if (v1 == 0 && v6 == 'x'.code) sb.append(u.get('e'.code)) // x → এক্স
                    v8 = v5; v5 = 0
                    if (v6 == 'w'.code && isConsonant(v1) && isConsonant(v9)) {
                        sb.append("্"); v8 = 0; v5 = 1 // w = ব-phala inside a cluster
                    }
                    if (v12 == 'k'.code && v9 == 'S'.code && v1 == 'h'.code &&
                        (v6 == 'N'.code || v6 == 'm'.code)
                    ) {
                        sb.append("্"); v8 = 0; v5 = 1 // ক্ষ + ণ/ম
                    }
                    sb.append(u.get(v6))
                }
            } else {
                // consonant + consonant
                if ((v6 == 'y'.code || v6 == 'Z'.code) &&
                    !(v6 == 'y'.code && v1 == 'q'.code && v9 == 'q'.code)
                ) v6 = 'z'.code
                if (v9 == 'k'.code && v1 == 'k'.code && v6 == 'h'.code) v1 = 'S'.code // kkh → ক্ষ

                val v3 = u.getDual(v6, v1)
                if (v3 == null || v7 != 0) {
                    v7 = 0; v8 = v5; v5 = 0
                    if (v9 == 'r'.code || v1 != 'r'.code || v6 != 'z'.code) {
                        if ((v1 != 'r'.code || v9 == 'r'.code) &&
                            (v1 != 'r'.code || v9 != 'r'.code || !isConsonant(v12))
                        ) {
                            if (v1 != 'r'.code || v9 != 'r'.code || (!isVowel(v12) && v12 != 0)) {
                                if (!notJukta(v12, v9, v1, v6)) {
                                    sb.append("্")
                                    v5 = 1
                                }
                            } else {
                                // rr after vowel/word start → reph: drop the 2nd র, add ্.
                                sb.delete(sb.length - 1, sb.length)
                                sb.append("্")
                            }
                        }
                    } else {
                        sb.append("‍্") // r+z → র‍্য (ZWJ keeps the ya-phala form)
                    }
                    sb.append(u.get(v6))
                } else {
                    // Two-char consonant (kh, ch, Th…): replaces its first char.
                    v7 = 1
                    if (v12 == 'g'.code && v9 == 'k'.code && v1 == 'S'.code && v6 == 'h'.code) {
                        v5 = 0; v8 = 0
                    }
                    val v4 = if (isVowel(v9) || v9 == 0 || v8 != 0) 1 else 0
                    if (!dualSitsUnder(v12, v9, v1, v6) || v4 != 0) {
                        if (v5 == 0) sb.delete(sb.length - 1, sb.length)
                        else sb.delete(sb.length - 2, sb.length)
                        sb.append(v3)
                        v8 = v5; v5 = 0
                    } else {
                        sb.delete(sb.length - 1, sb.length)
                        if (v9 == 'r'.code && v12 == 'r'.code) sb.delete(sb.length - 1, sb.length)
                        if (v5 == 0 && v9 != 0 && !isVowel(v9)) sb.append("্")
                        sb.append(v3)
                        v8 = v5; v5 = 1
                    }
                }
            }

            v12 = v9; v9 = v1; v1 = v6
        }
        return sb.toString()
    }
}
